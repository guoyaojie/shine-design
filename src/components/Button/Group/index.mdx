---
name: Button.Group 按钮组
route: /doc/buttonGroup
menu: button
---

import JsxParser from "react-jsx-parser";
import {
  MessageBox,
  InlineCode,
  Section,
  Preview,
  Playground,
  Code,
  PropsTable
} from "theme";
import { Button } from "shineDev";

# Group 按钮组

## 概述

待补充

## 代码示例

