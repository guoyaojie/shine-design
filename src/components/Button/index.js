/**
 * @Component Button
 * @Type 按钮
 * @Author 瞿龙俊 - qulongjun@shine.design
 * @Date 2019/5/12 09:35
 */
import _ from 'lodash';
import React, {PureComponent} from 'react';
import classNames from 'classnames';
import * as PropTypes from 'prop-types';
import {classPrefix} from 'variables';
import './style/index.scss';

class Button extends PureComponent {
  render() {
    const {label, type, bgColor, className, attributes, children} = this.props;

    /** 计算样式 */
    const classes = classNames(
      `${classPrefix}-btn`,
      {
        [`${classPrefix}-btn-${bgColor}`]: _.isString(bgColor),
      },
      className,
    );

    return (
      // eslint-disable-next-line react/button-has-type
      <button
        type={type}
        className={classes}
        {...attributes}
      >
        {label || children}
      </button>
    );
  }
}

Button.propTypes = {
  /** 按钮提示文案 */
  label: PropTypes.node,
  /** 按钮类型，支持普通按钮:button，提交按钮:submit，重置按钮:reset */
  type: PropTypes.oneOf(['button', 'submit', 'reset']),
  /** 按钮背景颜色 */
  bgColor: PropTypes.string,
  /** 按钮轮廓颜色 */
  outlineColor: PropTypes.string,
  /** 按钮渐变色，支持设置两个属性：from：起始颜色，to：终止颜色 */
  gradient: PropTypes.shape({
    from: PropTypes.string,
    to: PropTypes.string,
  }),
  /** 按钮类型尺寸，支持 default：默认尺寸 / small：小尺寸 / large：大尺寸 */
  size: PropTypes.oneOf(['default', 'small', 'large']),
  /** 按钮是否被激活 */
  isActive: PropTypes.bool,
  /** 按钮是否被禁用 */
  isDisabled: PropTypes.bool,
  /** 用户自定义修饰符 */
  className: PropTypes.string,
  /** 用户自定义属性 */
  attributes: PropTypes.object,
};

Button.defaultProps = {
  label: 'Button',
  type: 'button',
  bgColor: 'primary',
  outlineColor: undefined,
  gradient: undefined,
  size: 'default',
  isActive: false,
  isDisabled: false,
  className: '',
  attributes: {},
};

export default Button;
