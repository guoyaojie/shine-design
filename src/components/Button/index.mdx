---
name: Button 按钮
route: /guide/button
menu: button
---

import JsxParser from "react-jsx-parser";
import {
  MessageBox,
  InlineCode,
  Example,
  Code,
  PropsTable
} from "theme";
import { Button } from "shineDev";
import code from "./demo";
import './demo/index.scss';

# Button 按钮

## 概述

用于触发若干操作命令或用户行为。

## 代码示例

#### 基础按钮
基于[Color 色彩](./color)配置，我们提供了所有色彩样式的基础按钮以及基础轮廓按钮。

<Example 
  display={<JsxParser {...code.basic} />}
  code={{code:code.basic.jsx}}
  components={code.basic.components}
/>

#### 按钮样式
针对不同风格的主题，提供了不同的边框样式可供选择，目前支持圆角矩形、直角矩形和椭圆矩形。

<Example 
  display={<JsxParser {...code.btnStyle} />}
  code={{code:code.btnStyle.jsx}}
  components={code.btnStyle.components}
/>

#### 按钮尺寸
组件提供除了默认值以外的两种不同尺寸，可以在不同场景下选择合适的按钮尺寸。

<Example 
  display={<JsxParser {...code.btnSize} />}
  code={{code:code.btnSize.jsx}}
  components={code.btnSize.components}
/>

### 配置参数

<PropsTable of={Button} />

