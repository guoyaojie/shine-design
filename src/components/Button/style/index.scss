/**
 * @Author 瞿龙俊 - qulongjun@shine.design
 * @Date 2018/9/5 下午5:38
 */

/**
 * 核心样式库
 */

//@import "../../../sass/mixins/rounded-mode";
//@import "../../../sass/mixins/icon";

/**
 * 依赖样式库
 */
//@import "~bootstrap/scss/functions";
//@import "~bootstrap/scss/variables";
//@import "~bootstrap/scss/mixins";
//@import "~bootstrap/scss/buttons";
//@import "~bootstrap/scss/button-group";
@import "~style/bootstrap/button/index";

//@import "../../../sass/mixins/button";
//@import "../../../sass/mixins/gradient";

/**
  独立样式
 */
//@import "./button";
