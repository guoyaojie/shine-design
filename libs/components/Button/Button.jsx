/**
 * @Component 按钮 - Button
 * @Type 基础组件
 * @Author 瞿龙俊 - qulongjun@shine.design
 * @Date 2018/9/5 下午1:48
 */

// 核心依赖
import React, {Component, Fragment} from 'react';
import * as PropTypes from 'prop-types';
// 第三方依赖库
import classNames from 'classnames'
import _ from 'lodash';
import colorType from '../../config/color';
// 样式
import './style';

// 组件依赖
import Icon from '../Icon/Icon';

export default class extends Component {
    constructor(props) {
        super(props);
    }

    static propTypes = {
        type: PropTypes.oneOf(['submit', 'button', 'reset']),
        btnStyle: PropTypes.oneOf(['square', 'pill', 'normal']),
        color: PropTypes.oneOf(_.concat(colorType, 'normal', colorType.map(color => `outline-${color}`))),
        gradient: PropTypes.shape({
            from: PropTypes.oneOf(colorType),
            to: PropTypes.oneOf(colorType)
        }),
        size: PropTypes.oneOf(['lg', 'sm', 'normal']),
        iconName: PropTypes.string,
        isWide: PropTypes.bool,
        isBlock: PropTypes.bool,
        isActive: PropTypes.bool,
        isDisabled: PropTypes.bool,
        isAir: PropTypes.bool,
        isRetina: PropTypes.bool,
        callbacks: PropTypes.object,
        attributes: PropTypes.object,
    };

    static defaultProps = {
        type: 'button',
        color: 'default',
        btnStyle: 'normal',
        size: 'normal',
        isWide: false,
        isBlock: false,
        isActive: false,
        isDisabled: false,
        isAir: false,
        isRetina: false
    };

    render() {
        const {type, value, btnStyle, color, gradient, size, isWide, isBlock, isActive, isDisabled, isAir, isRetina, iconName, callbacks, className, attributes} = this.props;
        const {from, to} = gradient || {};

        return (
            <Fragment>
                <button
                    type={type}
                    {...callbacks}
                    {...attributes}
                    className={classNames(
                        'btn', 'sh-btn',
                        {
                            [`btn-${color}`]: _.isString(color),
                            [`sh-btn--gradient-from-${from || ''}`]: _.isString(from) && _.isString(to),
                            [`sh-btn--gradient-to-${to || ''}`]: _.isString(from) && _.isString(to),
                            [`sh-btn--${btnStyle}`]: _.isString(btnStyle),
                            [`btn-${size}`]: _.isString(size) && !_.isEqual(size, 'normal'),
                            'sh-btn--icon': _.isString(iconName),
                            'sh-btn--wide': isWide,
                            'sh-btn--air': isAir,
                            'sh-btn--outline-2x': isRetina,
                            'btn-block': isBlock,
                            'active': isActive,
                            'disabled': isDisabled
                        },
                        ...(
                            _.isArray(className) ? className : [className]
                        )
                    )}>
                    {_.isString(iconName) ? (
                        <Fragment>
                            {
                                _.isUndefined(value) ? (
                                    <Icon className={iconName}/>
                                ) : (
                                    <span>
                                        <Icon className={iconName}/>
                                        <span>{value}</span>
                                    </span>
                                )
                            }
                        </Fragment>
                    ) : value}
                </button>
            </Fragment>
        );
    }
}