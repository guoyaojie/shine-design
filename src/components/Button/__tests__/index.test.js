import React from 'react';
import {render} from 'enzyme';
import Button from '../index';

describe('Button', () => {
    it('renders correctly', () => {
        const wrapper = render(<Button>Follow</Button>);
        expect(wrapper).toMatchSnapshot();
    });
});