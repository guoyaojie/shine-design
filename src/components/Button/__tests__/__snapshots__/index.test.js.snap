// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "type": "button",
    },
    "children": Array [
      Object {
        "data": "aaa",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
